### PR TITLE
[inductor] config to control whether we assume inputs are aligned

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122158

**Motivation**: https://github.com/pytorch/pytorch/issues/112771

**Summary**: Inductor generates triton that assumes that inputs are going to be 16-byte aligned. If the inputs aren't aligned, Inductor clones the inputs. This PR introduces a config option to not do this: when assume_aligned_inputs=False, Inductor will _not_ pass inputs as being divisible_by_16, and Inductor will not make clones. This an can generate code that might be a bit slower, but this tradeoff can be worth it in some scenarios where you might otherwise make a lot of clones.

Ideally, we could do this on a per-tensor basis. But this would be a lot of work, and attempts to add guards on storage offsets to do this automatically have run into issues: recompilations and excessive time to generate/check guards.

**Tests** https://github.com/pytorch/pytorch/pull/122159 flips this to False. It didn't run through all errors, but the ones we see are all expected failures: divisible_by_16 changes; triton kernel caching fails if we call the same triton kernel multiple times (this makes sense because the first call will have unaligned inputs, but subsequent calls have aligned inputs); and some xfailed tests start passing.

**Alternatives/RFC**:
* Is this the right thing to do with cudagraphs?
* Elias and Jason mentioned that we probably still want to make clones if we're dealing with unaligned inputs to matmuls. Is this something we should add in this config option? (In the use case I'm targeting, it seems like we don't need this optimization right now)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D55079094](https://our.internmc.facebook.com/intern/diff/D55079094)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55079094).

**Review by ezyang — APPROVED** [review]
Very nice, we should figure out a policy but this is a great start

**Review by eellison — COMMENTED** [review]
Nice ! We should be able to assume alignment on parameters / static_inputs. If folks are going to be testing perf on this internally might be worth doing ?

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55079094).

**Comment by davidberard98** [comment]
@eellison in the internal target this is apparently showing sufficiently good QPS. I think it’s reasonable to leave this as is until we have a use case where we expect the additional considerations to be necessary for QPS?

File: torch/_inductor/config.py [hunk 1]
```diff
@@ -463,6 +463,11 @@ def decide_compile_threads():
 # decompose some memory bound matmul/bmm to mul
 decompose_mem_bound_mm: bool = False
 
+# assume_aligned_inputs means that we assume that inputs will be aligned; we generate
+# code using this assumption, and clone tensors before use if they aren't aligned.
+# In the common case, most inputs will be aligned.
+assume_aligned_inputs: bool = True
```
Comment (gchanan) [inline torch/_inductor/config.py:469]: I'm a little confused by the naming.  If we are cloning unaligned tensors, aren't we _forcing_ the tensors to be aligned, not _assuming_ they are going to be aligned?  Or is this an inductor vs rest of the system perspective thing?

Comment (davidberard98) [reply torch/_inductor/config.py:469]: Whoops, didn't see this comment until now.

I guess my thought process was that, when this flag is turned on, inductor generates triton code that **assumes its inputs will be aligned**.

But I can see why this is confusing - do you have a suggestion for a better name?

Comment (davidberard98) [reply torch/_inductor/config.py:469]: or, put another way, assume_aligned_inputs=True -> optimize for the case where inputs are aligned, assume_aligned_inputs=False -> optimize for the case where inputs are unaligned

**Pull request closed without merging**